airbyte-cdk/python/airbyte_cdk/sources/streams/core.py

```diff
@@ -123,22 +142,25 @@ def read(  # type: ignore  # ignoring typing for ConnectorStateManager because o
         sync_mode = configured_stream.sync_mode
         cursor_field = configured_stream.cursor_field
 
-        slices = self.stream_slices(
-            cursor_field=cursor_field,
-            sync_mode=sync_mode,  # todo: change this interface to no longer rely on sync_mode for behavior
-            stream_state=stream_state,
+        # Get state from the stream's state getter or use the stream_state from incoming parameter if
+        # the connector uses legacy state
+        try:
+            stream_state = self.state  # type: ignore # we know the field might not exist...
```

Comment:
This was a tricky edge case to note. Some CDKs like file-based ones as part of the `self.state` setter will actually manipulate the state that gets set onto cursor. So rather than using the incoming `stream_state` parameter coming in, we should attempt to use the actual value stored on the stream's state. This was caught in tests.

For example for a file based connector:
Incoming State:
```
{
  "history": {
      "b.csv": "2023-06-05T03:54:07.000000Z",
      "c.csv": "2023-06-05T03:54:07.000000Z",
      "d.csv": "2023-06-05T03:54:07.000000Z",
}
```

When first setting state for the stream, we call `self.state` > `self._cursor.set_initial_state(value)`. This in turns sets state to:
```
{
    "history": {
        "b.csv": "2023-06-05T03:54:07.000000Z",
        "c.csv": "2023-06-05T03:54:07.000000Z",
        "d.csv": "2023-06-05T03:54:07.000000Z",
    },
    "_ab_source_file_last_modified": "2023-06-05T03:54:07.000000Z_d.csv",
}
```

So the last modified value is also set just by setting the state for the first time. We should try to avoid using the `stream_state` parameter whenever possible.

Reply:
yep I can add a permanent comment about why we need this